### ISSUE TITLE
Accessibility Recommendation

### ISSUE THREAD

**Issue by joe-watkins** [jonsuh/hamburgers#2]
Hello :) Good work here!

I'd love for you to consider users that may be using a keyboard or assistive technology. They can't send focus to the burger because `<divs>` are not focusable. This is because a `<div>` is used where perhaps `<a>` or `<button>` could be used. Doing this along with visually hiding some text content inside the anchor provides assistive technology something to work with.

You can test this by trying to access your burger by unplugging your mouse and try to tab to it. Here's a code snippet that could be considered:

```
<a class="hamburger" href="#some-navigation">
  <span class="visually-hidden">Main Navigation</span>
  <div class="hamburger-box">
    <div class="hamburger-inner"></div>
  </div>
</a>
```

`href` points to ID placed on element it controls. The author then could add `aria` to further enhance the accessibility. e.g. `aria-controls` `aria-expanded` etc.

If you MUST use a `<div>` adding `tabindex="0"` would make them focusable but is not ideal.  Also `role="button"` should be added as well. :(

**Comment by jonsuh** [comment]
Hey @joe-watkins, appreciate your kind words as well as the suggestion!

Accessibility is definitely something I consider a priority. I made sure to include a section about accessibility (further down in the README after customization) when I released it—it touches on accessibility and ARIA along with accompanying links to resources that goes more in depth.

I had considered wrapping it in an `<a>` but didn’t like the solution because semantically it wasn’t an anchor, and nesting `<div>`s in a `<button>` isn’t valid HTML5, which was why I resorted to `<div>` with `role="button"`.

However, you bring up an excellent point about `tabindex="0"` which I completely overlooked, so thanks for bringing that to my attention. For now I added `tabindex="0"` to the HTML example within accessibility. I’ll take some time in the next days to draft an explanation of its significance and importance within the section.

**Comment by gustavpursche** [comment]
@jonsuh @joe-watkins An `<a>` wouldn't be the right choice neither in this case. If there is an `href=""` it's a link - otherwise not. So it's a `<button>`. The sub-elements could be just spans. I don't really see the point in using a `<div>` instead. 

Every of those lovely hamburger should contain an invisible label in the end. You can achieve this by putting an invisible, but [screen-reader readable]( https://github.com/h5bp/html5-boilerplate/blob/5.3.0/dist/doc/css.md#visuallyhidden) element inside. Downside of this would be, the more complicated markup, and the extra class, you have to copy, if somebody wants to use the burgers. So what could help? Maybe `aria-label="Toggle Menu"` - see in [MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute).

Anyway - if you need help in refactoring/ planning, I'd love to hear from you. I'm so happy that you collected those patterns, since the designers around me were always looking for those :smiley_cat:

**Comment by jonsuh** [comment]
@gustavpursche Ah, didn’t realize nesting `<span>`s in a `<button>` is valid! Nice. I’ll work on adding support for markup that looks like the following:

```
<button class="hamburger" type="button">
  <span class="hamburger-box">
    <span class="hamburger-inner"></span>
  </span>
</button>
```

The section in the README also mentions ARIA and the example does use `aria-label`. Since it looks like people are overlooking it, I think I’ll add a link in the primary example about accessibility, which anchors down to the section on Accessibility.

As far as the `<input>`, I think it’s useful but only within `<form>`s. In the context of a hamburger icon, I think it doesn’t apply (since it’s unlikely the hamburger icon will be inside a form), but your `<button>` with nested `<span>`s suggestion is great!

**Comment by jonsuh** [comment]
Update: Just pushed a new release to support `<button>` markup and updated the README section on accessibility :+1:

**Comment by joe-watkins** [comment]
@jonsuh Thanks for having an open mind and considering #a11y :) I button is definitely a step in the right direction.

@gustavpursche These days, I actually prefer an `<a>` for Progressive Enhancement reasoning. Usually 'hamburgers' are triggering some sort of off-screen nav or something.  What if JavaScript is not available or not served to the user.. a `<button>` would fail. While it gives you more usability like the space and enter it will fail if no JS is a reality.

If you leverage an anchor tag you can take advantage of `:target` selectors! That is where the #ID comes in. If no JavaScript is present and the user interacts with anchor tag the container you have targeted can become visible and believe it or not the user's focus is sent to that container which is generally what you want as well for accessibility.

Here is an example:

```
#off-screen-nav {
  transform: translate(-100%); // offscreen
}

#off-screen-nav:target {
  transform: translate(0); // onscreen
}
```

The above code can't be taken advantage of with just a button.

Thoughts?

**Comment by jonsuh** [comment]
@joe-watkins `:target` selectors do wonders, totally agree, but I think that may extend beyond this project.

It’s difficult to guess the context of how these hamburger icons will be used. In the event JavaScript isn’t available, the user could just hide the menu button and make the menu visible with CSS. I thought it was responsible to suggest a better alternative to `<div>`s, but at some point I can only point users to resources on accessibility so they can better find solution that fits their context.

**Comment by joe-watkins** [comment]
@jonsuh I agree you could go down a rabbit hole huh :)

@gustavpursche Thanks for contributing to the conversation and helping getting these things focusable. :)

@jonsuh great point about context! `<button>` does the trick!
